bitcoin/src/bip32.rs

```diff
@@ -111,55 +122,136 @@ pub struct Xpub {
 #[cfg(feature = "serde")]
 internals::serde_string_impl!(Xpub, "a BIP-32 extended public key");
 
-/// A child number for a derived key
+/// A BIP32 types of key derivation
 #[derive(Copy, Clone, PartialEq, Eq, Debug, PartialOrd, Ord, Hash)]
 pub enum ChildNumber {
-    /// Non-hardened key
-    Normal {
-        /// Key index, within [0, 2^31 - 1]
-        index: u32,
-    },
-    /// Hardened key
-    Hardened {
-        /// Key index, within [0, 2^31 - 1]
-        index: u32,
-    },
+    /// Enum wrapper for a normal child number
+    Normal(NormalChildNumber),
+    /// Enum wrapper for a hardened child number
+    Hardened(HardenedChildNumber),
+}
+///Sub-module for HardenedChildNumber type
+mod hardened {
+    use super::*;
+    /// A child number for a derived key
+    /// Indexes 2147483648 to 4294967295 (the second half of all possible children)
+    #[derive(Copy, Clone, PartialEq, Eq, Debug, PartialOrd, Ord, Hash)]
+    pub struct HardenedChildNumber(u32);
+
+    impl HardenedChildNumber {
+        /// `HardenedChildNumber` with index 0.
+        pub const ZERO: Self = HardenedChildNumber(0);
+        /// `HardenedChildNumber` with index 1.
+        pub const ONE: Self = HardenedChildNumber(1);
+        /// HardenedChildNumber to u32 index
+        pub fn to_raw_number(self) -> u32 { self.0 }
+        /// HardenedChildNumber from u32 index
+        pub fn from_index(index: u32) -> Result<HardenedChildNumber, Error> {
+            if index & (1 << 31) == 0 {
+                Ok(HardenedChildNumber(index))
+            } else {
+                Err(Error::InvalidChildNumber(index))
+            }
+        }
+        /// HardenedChildNumber from ChildNumber
+        pub fn from_number(number: ChildNumber) -> Result<HardenedChildNumber, Error> {
```

Comment:
In b80658ed8c3955883284fe93699825cbed685a0d:

I guess you also want `HardenedChildNumber::to_number`.

I would also suggest making all the methods on `HardenedChildNumber` and `NormalChildNumber` `const` since they're all simple conversions.

Reply:
Good Idea